### PR TITLE
feat/AB#104447_EIS-configure-file-types-max-size-and-number-of-files

### PR DESCRIPTION
**Pull request by unai-reliefapp**
# Description

feat: add custom component for accepted files dropdown for file question types
feat: integrate max number of files allowed
feat: make max size check not default

## Useful links

- [Link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/104447)
- Please insert link to back-end branch if any
- Please insert any useful link ( documentation you used for example )

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Screenshots

New accepted types dropdown
https://github.com/user-attachments/assets/0acae219-9e7d-4254-86d0-21490afdf136

On multiple file check
I set the minimum file amount to 2(multiple file, at least 2, if not user can unselect the allow multiple option)
image
image

# Checklist:

( * == Mandatory ) 

- [X] * I have set myself as assignee of the pull request
- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

### CODE REVIEW

**Comment by AntoineRelief** [comment]
@unai-reliefapp 
I'm just wondering if we couldn't better indicate the number of files for users
like, can they see the maximum number of files + the max file size?

as it is appearing in the web element I put in doc-upload component

**Comment by unai-reliefapp** [comment]
image
If no limitation set, text wont be shown, the bytes value should be mapped to another format?
cc @AntoineRelief

**Comment by AntoineRelief** [comment]
@unai-reliefapp 
I think we should always have a limitation, so make it mandatory

**Comment by unai-reliefapp** [comment]
> @unai-reliefapp I think we should always have a limitation, so make it mandatory

File size limitation is always there, it would always display, but the file number no, if no multiple allow is selected right?
Anyway the styling as description is fine?
I'll use survey location to add the message in english and french

**Comment by AntoineRelief** [comment]
> > @unai-reliefapp I think we should always have a limitation, so make it mandatory
> 
> File size limitation is always there, it would always display, but the file number no, if no multiple allow is selected right? Anyway the styling as description is fine? I'll use survey location to add the message in english and french

@unai-reliefapp 
indeed, number of files should be mandatory only is multiple files are allowed.

If possible, I'd put the text above the button & below the text for drag & drop, but I don't know how complex it would be to do that

**Comment by unai-reliefapp** [comment]
@AntoineRelief 
image

I had many troubles to update the drag placeholder before the render, i manage to make it work with a generic event on creating the survey.
The message wont show in the form builder, but i think it's fine because user has access to the properties
Should I push the changes?

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

that's fine for me
does it display in the preview?

**Comment by unai-reliefapp** [comment]
> @unai-reliefapp
> 
> that's fine for me does it display in the preview?

yes @AntoineRelief 
image

**Comment by AntoineRelief** [comment]
@unai-reliefapp 
then push please

File: libs/shared/src/lib/components/form-builder/form-builder.component.ts [hunk 1]
```diff
@@ -1,35 +1,36 @@
+import { Dialog } from '@angular/cdk/dialog';
```
Comment (unai-reliefapp) [inline libs/shared/src/lib/components/form-builder/form-builder.component.ts:1]: Changes in imports are just a reorder @AntoineRelief

**Comment by unai-reliefapp** [comment]
all pushed @AntoineRelief

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

two things about this one:
- the limit is not working. It's only working if you upload all files at the same time:
Capture d’écran 2024-10-22 à 16 35 39
Capture d’écran 2024-10-22 à 16 35 34

- the display of bytes is not helping: 
Capture d’écran 2024-10-22 à 16 36 32
you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:

Capture d’écran 2024-10-22 à 16 37 35

Putting back as draft

**Comment by unai-reliefapp** [comment]
> @unai-reliefapp
> 
> two things about this one:
> 
> * the limit is not working. It's only working if you upload all files at the same time:
>   Capture d’écran 2024-10-22 à 16 35 39
>   Capture d’écran 2024-10-22 à 16 35 34
> * the display of bytes is not helping:
>   Capture d’écran 2024-10-22 à 16 36 32
>   you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:
> 
> Capture d’écran 2024-10-22 à 16 37 35
> 
> Putting back as draft

So from the initial capture I sent you in the previous message,

- Should I convert it to MB and enough or reproduce the whole message used in their doc upload component?
- Limit is not working what do you mean exactly? Can you provide me a more specific example? You mean that if i upload a file with less size, and then upload one with higher size, limitation is not triggering?

**Comment by AntoineRelief** [comment]
> > @unai-reliefapp
> > two things about this one:
> > 
> > * the limit is not working. It's only working if you upload all files at the same time:
> >   Capture d’écran 2024-10-22 à 16 35 39
> >   Capture d’écran 2024-10-22 à 16 35 34
> > * the display of bytes is not helping:
> >   Capture d’écran 2024-10-22 à 16 36 32
> >   you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:
> > 
> > Capture d’écran 2024-10-22 à 16 37 35
> > Putting back as draft
> 
> So from the initial capture I sent you in the previous message,
> 
> * Should I convert it to MB and enough or reproduce the whole message used in their doc upload component?

Just the conversion to MB ( or anything else, based on the size, it could be KB for example )

> * Limit is not working what do you mean exactly? Can you provide me a more specific example? You mean that if i upload a file with less size, and then upload one with higher size, limitation is not triggering?

- set the limit to 3
- click on button and upload 3 files
- click on the button to add other files to the question, and select another one
- the 4th file is correctly added to the question ( as per screenshot )

**Comment by unai-reliefapp** [comment]
> > > @unai-reliefapp
> > > two things about this one:
> > > 
> > > * the limit is not working. It's only working if you upload all files at the same time:
> > >   Capture d’écran 2024-10-22 à 16 35 39
> > >   Capture d’écran 2024-10-22 à 16 35 34
> > > * the display of bytes is not helping:
> > >   Capture d’écran 2024-10-22 à 16 36 32
> > >   you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:
> > > 
> > > Capture d’écran 2024-10-22 à 16 37 35
> > > Putting back as draft
> > 
> > 
> > So from the initial capture I sent you in the previous message,
> > 
> > * Should I convert it to MB and enough or reproduce the whole message used in their doc upload component?
> 
> Just the conversion to MB ( or anything else, based on the size, it could be KB for example )
> 
> > * Limit is not working what do you mean exactly? Can you provide me a more specific example? You mean that if i upload a file with less size, and then upload one with higher size, limitation is not triggering?
> 
> * set the limit to 3
> * click on button and upload 3 files
> * click on the button to add other files to the question, and select another one
> * the 4th file is correctly added to the question ( as per screenshot )

Fixes pushed 👍 @AntoineRelief 

image


https://github.com/user-attachments/assets/309e0be4-7813-4991-a2d1-705b8c2d98b8

**Comment by AntoineRelief** [comment]
> > > > @unai-reliefapp
> > > > two things about this one:
> > > > 
> > > > * the limit is not working. It's only working if you upload all files at the same time:
> > > >   Capture d’écran 2024-10-22 à 16 35 39
> > > >   Capture d’écran 2024-10-22 à 16 35 34
> > > > * the display of bytes is not helping:
> > > >   Capture d’écran 2024-10-22 à 16 36 32
> > > >   you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:
> > > > 
> > > > Capture d’écran 2024-10-22 à 16 37 35
> > > > Putting back as draft
> > > 
> > > 
> > > So from the initial capture I sent you in the previous message,
> > > 
> > > * Should I convert it to MB and enough or reproduce the whole message used in their doc upload component?
> > 
> > Just the conversion to MB ( or anything else, based on the size, it could be KB for example )
> > 
> > > * Limit is not working what do you mean exactly? Can you provide me a more specific example? You mean that if i upload a file with less size, and then upload one with higher size, limitation is not triggering?
> > 
> > * set the limit to 3
> > * click on button and upload 3 files
> > * click on the button to add other files to the question, and select another one
> > * the 4th file is correctly added to the question ( as per screenshot )
> 
> Fixes pushed 👍 
> 
> image
> 

Quick as hell
Just to be sure, as I think I was not clear
If the value can be converted to MB ( bigger than 1MB ), display MB. Else, display KB

**Comment by unai-reliefapp** [comment]
> > > > > @unai-reliefapp
> > > > > two things about this one:
> > > > > 
> > > > > * the limit is not working. It's only working if you upload all files at the same time:
> > > > >   Capture d’écran 2024-10-22 à 16 35 39
> > > > >   Capture d’écran 2024-10-22 à 16 35 34
> > > > > * the display of bytes is not helping:
> > > > >   Capture d’écran 2024-10-22 à 16 36 32
> > > > >   you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:
> > > > > 
> > > > > Capture d’écran 2024-10-22 à 16 37 35
> > > > > Putting back as draft
> > > > 
> > > > 
> > > > So from the initial capture I sent you in the previous message,
> > > > 
> > > > * Should I convert it to MB and enough or reproduce the whole message used in their doc upload component?
> > > 
> > > 
> > > Just the conversion to MB ( or anything else, based on the size, it could be KB for example )
> > > > * Limit is not working what do you mean exactly? Can you provide me a more specific example? You mean that if i upload a file with less size, and then upload one with higher size, limitation is not triggering?
> > > 
> > > 
> > > 
> > > * set the limit to 3
> > > * click on button and upload 3 files
> > > * click on the button to add other files to the question, and select another one
> > > * the 4th file is correctly added to the question ( as per screenshot )
> > 
> > 
> > Fixes pushed 👍
> > image
> 
> Quick as hell Just to be sure, as I think I was not clear If the value can be converted to MB ( bigger than 1MB ), display MB. Else, display KB

on going

**Comment by unai-reliefapp** [comment]
> > > > > > @unai-reliefapp
> > > > > > two things about this one:
> > > > > > 
> > > > > > * the limit is not working. It's only working if you upload all files at the same time:
> > > > > >   Capture d’écran 2024-10-22 à 16 35 39
> > > > > >   Capture d’écran 2024-10-22 à 16 35 34
> > > > > > * the display of bytes is not helping:
> > > > > >   Capture d’écran 2024-10-22 à 16 36 32
> > > > > >   you should convert that to a round number ( like, 7MB, etc ). It can remain as it is in the settings, but for end-users, this is not correct. I prefer how it's done in the web element for example:
> > > > > > 
> > > > > > Capture d’écran 2024-10-22 à 16 37 35
> > > > > > Putting back as draft
> > > > > 
> > > > > 
> > > > > So from the initial capture I sent you in the previous message,
> > > > > 
> > > > > * Should I convert it to MB and enough or reproduce the whole message used in their doc upload component?
> > > > 
> > > > 
> > > > Just the conversion to MB ( or anything else, based on the size, it could be KB for example )
> > > > > * Limit is not working what do you mean exactly? Can you provide me a more specific example? You mean that if i upload a file with less size, and then upload one with higher size, limitation is not triggering?
> > > > 
> > > > 
> > > > 
> > > > * set the limit to 3
> > > > * click on button and upload 3 files
> > > > * click on the button to add other files to the question, and select another one
> > > > * the 4th file is correctly added to the question ( as per screenshot )
> > > 
> > > 
> > > Fixes pushed 👍
> > > image
> > 
> > 
> > Quick as hell Just to be sure, as I think I was not clear If the value can be converted to MB ( bigger than 1MB ), display MB. Else, display KB
> 
> on going

image
Correct format based on size also pushed 👍

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

I found two issues, perhaps you should start with that, before you work on expressions:
- text would display number of files even if single upload ( already fixed, see last commit )
- still an issue with upload of files in multi upload question. Create a question with for example 3 files maximum, then display the form to create a new record. Attach 3 files, remove one, and try to upload a new one. You will see the error message. Indeed, the temporary file storage is not cleared when we remove a file, which explains the issue. There is certainly an event from surveyjs, similar to onUpload / onDownload that we use to build the temporary file storage, that we should listen on

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

main suspect:

`  private onClearFiles(options: any): void {
    options.callback('success');
  }`

**Comment by unai-reliefapp** [comment]
clearFiles call back implementation pushed @AntoineRelief

File: libs/shared/src/lib/services/form-builder/form-builder.service.ts [hunk 2]
```diff
@@ -216,9 +218,11 @@ export class FormBuilderService {
   /**
    * Handles the clearing of files
    *
+   * @param temporaryFilesStorage Temporary files saved while executing the survey
    * @param options Options regarding the files
    */
-  private onClearFiles(options: any): void {
+  private onClearFiles(temporaryFilesStorage: any, options: any): void {
+    temporaryFilesStorage[options.name] = options.value;
```
Comment (AntoineRelief) [inline libs/shared/src/lib/services/form-builder/form-builder.service.ts]: @unai-reliefapp 

to test ( I haven't tested it )
what if you upload some files, then save the record, then open it again
how does it behave?

does it work fine?

Comment (unai-reliefapp) [reply libs/shared/src/lib/services/form-builder/form-builder.service.ts]: there were multiple checks and error handler incorrectly implemented or missing
Now all cases are covered @AntoineRelief 👍 

https://github.com/user-attachments/assets/79dcf808-c0a2-4d85-9fc1-e56dcfdc3dfc

File: libs/shared/src/lib/survey/components/accepted-value-types-text/accepted-value-types-text.component.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  /** Form control for accepted value types */
+  public formGroup = this.fb.group({
+    valueTypes: [['']],
+  });
+
+  /** Destroy subject */
+  private destroy$: Subject<void> = new Subject<void>();
+
+  /**
+   * The constructor function is a special function that is called when a new instance of the class is
```
Comment (AntoineRelief) [inline libs/shared/src/lib/survey/components/accepted-value-types-text/accepted-value-types-text.component.ts]: @unai-reliefapp 

please no more of these jsdocs about constructors 😂

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

great work 👍 

I only found one issue, when creating a new record with multi files, adding multiple files in the question, then removing one, then adding one

that explains the changes I did in the last commit to fix that, the rest is good 👍

**Pull request merged**